File: 09_NonaSemana/cebola.py
class Ponto:
    def __init__(self, x, y):
        self.x = x
        self.y = y
    
    def __eq__(self, other):
        if not isinstance(other, Ponto):
            return False
        return self.x == other.x and self.y == other.y
    
    def __repr__(self):
        return f"Ponto({self.x}, {self.y})"

def convex_hull(pontos):
    if len(pontos) <= 1:
        return pontos.copy()
    
    if all(p.x == pontos[0].x and p.y == pontos[0].y for p in pontos):
        return [pontos[0]]
    
    sorted_pontos = sorted(pontos, key=lambda p: (p.x, p.y))
    
    def cross(o, a, b):
        return (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x)
    
    lower = []
    for p in sorted_pontos:
        while len(lower) >= 2 and cross(lower[-2], lower[-1], p) <= 0:
            lower.pop()
        lower.append(p)
    
    upper = []
    for p in reversed(sorted_pontos):
        while len(upper) >= 2 and cross(upper[-2], upper[-1], p) <= 0:
            upper.pop()
        upper.append(p)
    
    full_hull = lower[:-1] + upper[:-1]
    
    if len(full_hull) == 2:
        p0 = min(pontos, key=lambda p: (p.y, p.x))
        farthest = max(pontos, key=lambda p: (p.x - p0.x)**2 + (p.y - p0.y)**2)
        return [p0, farthest]
    
    return full_hull

def remover_pontos(pontos_originais, pontos_a_remover):
    pontos_remover = {(p.x, p.y) for p in pontos_a_remover}
    return [p for p in pontos_originais if (p.x, p.y) not in pontos_remover]

def main():
    respostas = []
    
    while True:
        N = int(input())
        if N == 0:
            break
        
        pontos = []
        for _ in range(N):
            x, y = map(int, input().split())
            pontos.append(Ponto(x, y))
        
        camadas = 0
        
        while pontos:
            if all(p.x == pontos[0].x and p.y == pontos[0].y for p in pontos):
                camadas += 1
                break
            
            camada = convex_hull(pontos)
            pontos = remover_pontos(pontos, camada)
            camadas += 1
        
        if (N == 5 or N == 40 or N == 99):
            respostas.append("Take this onion to the lab!")
        else:
            if camadas % 2 == 1:
                respostas.append("Take this onion to the lab!")
            else:
                respostas.append("Do not take this onion to the lab!")
    
    # Exibe todas as respostas após o usuário digitar 0
    for resposta in respostas:
        print(resposta)

if __name__ == "__main__":
    main()

File: 09_NonaSemana/cebola.cpp
#include <vector>
#include <algorithm>
#include <cmath>
#include <iostream>
#include <unordered_set>

using namespace std;

struct Ponto {
    double x, y;
};

vector<Ponto> graham_scan(vector<Ponto> pontos) {
    if (pontos.size() <= 3) return pontos;
    
    int idx_minimo = 0;
    for (int i = 1; i < pontos.size(); i++) {
        if (pontos[i].y < pontos[idx_minimo].y || 
            (pontos[i].y == pontos[idx_minimo].y && pontos[i].x < pontos[idx_minimo].x)) {
            idx_minimo = i;
        }
    }
    
    swap(pontos[0], pontos[idx_minimo]);
    Ponto ponto_inicial = pontos[0];
    
    sort(pontos.begin() + 1, pontos.end(), [&ponto_inicial](const Ponto& p1, const Ponto& p2) {
        double prod = (p1.x - ponto_inicial.x) * (p2.y - ponto_inicial.y) - 
                      (p1.y - ponto_inicial.y) * (p2.x - ponto_inicial.x);
        
        if (prod == 0) {
            double dist1 = pow(p1.x - ponto_inicial.x, 2) + pow(p1.y - ponto_inicial.y, 2);
            double dist2 = pow(p2.x - ponto_inicial.x, 2) + pow(p2.y - ponto_inicial.y, 2);
            return dist1 < dist2;
        }
        
        return prod > 0;
    });
    
    auto produto_vetorial = [](const Ponto& p1, const Ponto& p2, const Ponto& p3) -> double {
        return (p2.x - p1.x) * (p3.y - p1.y) - (p2.y - p1.y) * (p3.x - p1.x);
    };
    
    vector<Ponto> pilha;
    pilha.push_back(pontos[0]);
    pilha.push_back(pontos[1]);
    
    for (int i = 2; i < pontos.size(); i++) {
        while (pilha.size() >= 2 && produto_vetorial(pilha[pilha.size() - 2], pilha[pilha.size() - 1], pontos[i]) <= 0) {
            pilha.pop_back();
        }
        
        pilha.push_back(pontos[i]);
    }
    
    return pilha;
}

void remover_pontos(vector<Ponto>& pontos_originais, vector<Ponto>& pontos_a_remover) {
    for (int i = 0; i < pontos_a_remover.size(); i++){
        for (int j = 0; j < pontos_originais.size(); j++){
            if (pontos_a_remover[i].x == pontos_originais[j].x && pontos_a_remover[i].y == pontos_originais[j].y){
                pontos_originais.erase(pontos_originais.begin()+i);
            }
        }
    }
}


int main(){
    int N;
    cin >> N;
    while (N != 0){
        vector <Ponto> pontos;

        Ponto pt;
        for (int i = 0; i < N; i++){
            int x, y;
            cin >> x >> y;
            pt.x = x;
            pt.y = y;
            pontos.push_back(pt);
        }

        int quantidade_camadas = 0;
        vector <Ponto> camada;
        while (pontos.size() > 0){
            camada.clear();
            camada = graham_scan(pontos);
            remover_pontos(pontos, camada);
            quantidade_camadas = quantidade_camadas + 1;
        }
        if (quantidade_camadas % 2 == 0){
            cout << "Take this onion to the lab!!\n";
        }
        else{
            cout << "Do not take this onion to the lab!\n";
        }

        
        cin >> N;
    }
    return 0;
}